pkg/kubelet/cm/cpumanager/topology_hints.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpumanager
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/klog"
+	"math"
+
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/topology"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/socketmask"
+)
+
+func (m *manager) GetTopologyHints(pod v1.Pod, container v1.Container) []topologymanager.TopologyHint {
+	var cpuHints []topologymanager.TopologyHint
+	for resourceObj, amountObj := range container.Resources.Requests {
+		resource := string(resourceObj)
+		requested := int(amountObj.Value())
+		if resource != "cpu" {
+			continue
+		}
+
+		klog.Infof("[cpumanager] Guaranteed CPUs detected: %v", requested)
+
+		topo, err := topology.Discover(m.machineInfo)
+		if err != nil {
+			klog.Infof("[cpu manager] error discovering topology")
+		}
+		var assignableCPUs cpuset.CPUSet
+		containerID, _ := findContainerIDByName(&pod.Status, container.Name)
+		if cset, ok := m.state.GetCPUSet(containerID); ok {
+			klog.Infof("[cpumanager] Reusing pre-assigned CPUSet: %v", cset)
+			assignableCPUs = cset
+		} else {
+			// Otherwise, calculate the assignable CPUs from the topology.
+			assignableCPUs = m.getAssignableCPUs(topo)
+		}
+		klog.Infof("AssignableCPUs: %v", assignableCPUs)
+		cpuAccum := newCPUAccumulator(topo, assignableCPUs, requested)
+
+		//Can we always assume NumSockets and Socket Numbers are the same?
+		socketCount := topo.NumSockets
+		klog.Infof("[cpumanager] Number of sockets on machine (available and unavailable): %v", socketCount)
+		cpuHintsTemp := getCPUMask(socketCount, cpuAccum, requested)
+		CPUsPerSocket := topo.CPUsPerSocket()
+		cpuHints = getPreferred(cpuHintsTemp, requested, CPUsPerSocket)
+		klog.Infof("CPUHints: %v", cpuHints)
+	}
+	return cpuHints
+}
+
+func (m *manager) getAssignableCPUs(topo *topology.CPUTopology) cpuset.CPUSet {
```

Comment:
This feels like it should be more generally usable by the CPUManager as a whole. We should consider a cleanup PR in the future that consolidates this logic into the mainline CPUManager instead of duplicating it here.